azure/service.ts

```diff
@@ -21,25 +21,249 @@ import { RunError } from "@pulumi/pulumi/errors";
+        let targetPortNumber: number | undefined;
+        let protocol: string | undefined;
+        let isPublic: boolean | undefined;
+        if (ports) {
+            if (ports.length >= 2) {
+                throw new RunError("Only zero or one port can be provided with a container: " + containerName);
+            }
+
+            if (ports.length === 1) {
+                const port = ports[0];
+                hostPortNumber = port.port;
+                targetPortNumber = port.targetPort !== undefined ? port.targetPort : hostPortNumber;
+                protocol = port.protocol;
+
+                if (targetPortNumber !== hostPortNumber) {
+                    throw new RunError("Mapping a host port to a different target port is not supported in Azure currently.");
+                }
+
+                if (containerName === firstContainerName) {
+                    firstContainerPort = targetPortNumber;
+                }
+
+                const external = port.external === undefined ? false : port.external;
+                if (isPublic === undefined) {
+                    // first port we're seeing.  set the isPublic value based on that.
+                    isPublic = external;
+                }
+                else if (isPublic !== external) {
+                    // have an existing port.  Values have to match.
+                    throw new RunError("All ports must have a matching [external] value.");
+                }
+            }
+        }
+
+        if (isPublic === false) {
+            throw new RunError("Only public ip address types are supported by Azure currently.");
+        }
+
+        const imageName = getImageName(container);
+        if (!imageName) {
+            throw new Error("[getImageName] should have always produced an image name.");
+        }
+
+        const registry = container.build ? getOrCreateGlobalRegistry() : undefined;
+
+        const imageOptions = computeImage(this, imageName, container, registry);
+
+        const memoryInGB = pulumi.output(container.memoryReservation).apply(
+            r => r === undefined ? 1 : r / 1024);
+
+        const qualifiedName = (name + "-" + containerName).replace(disallowedChars, "-");
+        azureContainers.push({
+            name: qualifiedName,
+            cpu: 1,
+            memory: memoryInGB,
+            port: targetPortNumber,
+            protocol: protocol,
+            image: imageOptions.apply(io => io.image),
+            environmentVariables: imageOptions.apply(io => io.environment),
+        });
+
+        credentials = credentials || (registry
+            ? [{ password: registry.adminPassword, server: registry.loginServer, username: registry.adminUsername }]
+            : undefined);
+
+        if (targetPortNumber !== undefined) {
+            exposedPorts[containerName] = { [targetPortNumber]: hostPortNumber! };
+        }
+    }
+
+    const group = new azure.containerservice.Group(
```

Comment:
Are we just using AKS for this?

Frankly, for all the new work we do, I'd just assume Kubernetes and forget the rest of the legacy boneyard experiments that came before... 😉

Reply:
I also hope this will help with lame restrictions like "replicas == 1".

Reply:
Right now, we are not using Kubernetes at all here - we are building directly on Azure Container Instances (the Fargate-Iike feature on Azure).

However, there are several significant limitations to ACI even relative to Fargate, and it is generally not clear we can provide a rich notion of cloud.Service natively without the breadth of a Kubernetes-like orchestration underneath. 

If we go the path of assuming Kubernetes here - the whole notion of `cloud.Service` (And the cloud framework itself for that matter) becomes quite different.  This is a topic we need to have a deep discussion on now that we have a best-attempt at implementing this API on a second cloud in hand.

Reply:
Btw - the `azure.containerservice.Group` name was likely misleading here. We frankly got a little too opinionated (that is, wrong) with our namespacing in `@pulumi/azure`, which has lead to several places where our projected names don’t match the normal terminology used for the platform feature.  In this case, we combined several unrelated “container” features into a single namespace.

Reply:
Opened https://github.com/pulumi/pulumi-azure/issues/97 on this latter point.

Reply:
We need to just burn the ships and go all in non Kubernetes at this point. Let's discuss offline, but let's make sure we do so asap.